here-naksha-lib-psql/src/commonMain/kotlin/naksha/psql/executors/write/UpdateFeature.kt

```diff
@@ -0,0 +1,206 @@
+package naksha.psql.executors.write
+
+import naksha.base.Int64
+import naksha.model.Flags
+import naksha.model.Metadata
+import naksha.model.Metadata.Metadata_C.geoGrid
+import naksha.model.Metadata.Metadata_C.hash
+import naksha.model.NakshaError
+import naksha.model.NakshaException
+import naksha.model.Tuple
+import naksha.model.Version
+import naksha.model.objects.NakshaFeature
+import naksha.model.request.Write
+import naksha.psql.PgCollection
+import naksha.psql.PgColumn
+import naksha.psql.PgCursor
+import naksha.psql.PgSession
+import naksha.psql.PgUtil.PgUtilCompanion.quoteIdent
+import naksha.psql.executors.write.WriteFeatureUtils.newFeatureTupleNumber
+import naksha.psql.executors.write.WriteFeatureUtils.resolveFlags
+import naksha.psql.executors.write.WriteFeatureUtils.tuple
+
+class UpdateFeature(
+    val session: PgSession
+) {
+
+    fun execute(collection: PgCollection, write: Write): Tuple {
+        val feature = write.feature?.proxy(NakshaFeature::class) ?: throw NakshaException(
+            NakshaError.ILLEGAL_ARGUMENT,
+            "UPDATE without feature"
+        )
+        require(feature.id == write.featureId) {
+            "Feature id in payload (${feature.id}) and write request (${write.featureId}) are different"
+        }
+
+        val cursorWithFeatureHead = fetchFromHead(collection, feature.id)
+        val previousMetadata = metaFromCursor(cursorWithFeatureHead)
+
+        require(feature.id == previousMetadata.id) {
+            "Feature id (${feature.id}) differs from previous metadata (${previousMetadata.id})"
+        }
+        require(previousMetadata.nextVersion == null) {
+            "Previous metadata shouldn't have 'nextVersion' but it does (${previousMetadata.nextVersion})"
+        }
+
+        val tupleNumber = newFeatureTupleNumber(collection, feature.id, session)
+        val flags = resolveFlags(collection, session)
+        val newVersion = Version(previousMetadata.version.txn + 1)
+        val tuple = tuple(
+            session.storage,
+            tupleNumber,
+            feature,
+            metadataForNewVersion(previousMetadata, newVersion, feature, flags),
+            write.attachment,
+            flags
+        )
+
+        removeFeatureFromDel(collection, feature.id)
+        collection.history?.let { hstTable ->
+            insertPreviousVersionToHst(hstTable.name, cursorWithFeatureHead, newVersion)
+        }
+        cursorWithFeatureHead.close() // we don't need it anymore, TODO: refactor for nicer usage
```

Comment:
would it be closed in case of error line before?

Reply:
Nope, great catch. 
I added `close` right after the execution whenever cursor is not needed.
For when it's needed, I got something like this:
```
object PgCursorUtil {

    fun <T> PgCursor.collectAndClose(collector: (ReadOnlyRow) -> T): List<T> {
        val result = mutableListOf<T>()
        try {
            while (next()) {
                result.add(collector(ReadOnlyRow(this)))
            }
        } catch (e: Exception) {
            close()
            throw e
        }
        return result
    }

    class ReadOnlyRow(private val cursor: PgCursor) {
        fun column(column: PgColumn): Any? =
            cursor.column(column)

        operator fun <T : Any> get(column: PgColumn): T =
            cursor[column]
    }
}
```